opendevin/core/config.py

```diff
@@ -211,38 +375,55 @@ class SandboxConfig(metaclass=Singleton):
     od_runtime_startup_env_vars: dict[str, str] = field(default_factory=dict)
     browsergym_eval_env: str | None = None
 
-    def defaults_to_dict(self) -> dict:
-        """Serialize fields to a dict for the frontend, including type hints, defaults, and whether it's optional."""
-        dict = {}
-        for f in fields(self):
-            dict[f.name] = get_field_info(f)
-        return dict
-
-    def __str__(self):
-        attr_str = []
-        for f in fields(self):
-            attr_name = f.name
-            attr_value = getattr(self, f.name)
-
-            attr_str.append(f'{attr_name}={repr(attr_value)}')
-
-        return f"SandboxConfig({', '.join(attr_str)})"
+    @classmethod
+    def load_from_env(cls, env_dict: dict[str, str]) -> 'SandboxConfig':
+        config = cls()
+        for f in fields(cls):
+            env_var_name = f'SANDBOX_{f.name.upper()}'
+            if env_var_name in env_dict:
+                value = env_dict[env_var_name]
+                if value:
+                    setattr(config, f.name, cls._cast_value(f.type, value))
+        return config
+
+    @classmethod
+    def load_from_toml(cls, toml_config: dict) -> 'SandboxConfig':
+        sandbox_config = (
+            cls()
+        )  # This will either create a new instance or return the existing one
+
+        # First, migrate old sandbox configs from [core] section
+        if 'core' in toml_config:
+            core_config = toml_config['core']
+            keys_to_migrate = [key for key in core_config if key.startswith('sandbox_')]
+            for key in keys_to_migrate:
+                new_key = key.replace('sandbox_', '')
+                if hasattr(sandbox_config, new_key):
+                    setattr(sandbox_config, new_key, core_config[key])
+                else:
+                    logger.opendevin_logger.warning(f'Unknown sandbox config: {key}')
+
+        # Then, override with new-style [sandbox] section if it exists
+        if 'sandbox' in toml_config:
+            # Use the singleton's update mechanism
+            cls(**toml_config['sandbox'])
 
-    def __repr__(self):
-        return self.__str__()
+        return sandbox_config
 
 
 class UndefinedString(str, Enum):
     UNDEFINED = 'UNDEFINED'
 
 
 @dataclass
-class AppConfig(metaclass=Singleton):
+class AppConfig(BaseConfig, metaclass=SingletonABCMeta):
     """Configuration for the app.
 
     Attributes:
         llms: A dictionary of name -> LLM configuration. Default config is under 'llm' key.
+        memories: A dictionary of name -> Memory configuration. Default config is under 'memory' key.
         agents: A dictionary of name -> Agent configuration. Default config is under 'agent' key.
+        memories: A dictionary of name -> Memory configuration. Default config is under 'memory' key.
```

Comment:
duplicate

Reply:
Which one was a duplicate? `memories` is just one?

Reply:
Yes, the "memories" line was it.